src/writer.rs

```diff
@@ -0,0 +1,373 @@
+use num_traits::FromPrimitive;
+
+use {NiftiHeader, util::is_gz_file};
+
+type B = LittleEndian;
+
+/// Specific actions and header information that depend on the ndarray type.
+pub trait TypeTool<T> {
+
+    /// Returns the `datatype` mapped to the type T
+    fn to_data_type() -> i16;
+
+    /// Calls the right `write` method. For example, if T is i16, call `write_i16`.
+    fn write<W: WriteBytesExt>(bytes: &mut W, d: T);
+}
+
+impl TypeTool<u8> for u8 {
+    fn to_data_type() -> i16 { 2 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: u8)
+    {
+        writer.write_u8(d).unwrap();
+    }
+}
+impl TypeTool<i16> for i16 {
+    fn to_data_type() -> i16 { 4 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: i16) {
+        writer.write_i16::<B>(d).unwrap();
+    }
+}
+impl TypeTool<i32> for i32 {
+    fn to_data_type() -> i16 { 8 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: i32) {
+        writer.write_i32::<B>(d).unwrap();
+    }
+}
+impl TypeTool<f32> for f32 {
+    fn to_data_type() -> i16 { 16 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: f32) {
+        writer.write_f32::<B>(d).unwrap();
+    }
+}
+// NIFTI_TYPE_COMPLEX64      32
+impl TypeTool<f64> for f64 {
+    fn to_data_type() -> i16 { 64 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: f64) {
+        writer.write_f64::<B>(d).unwrap();
+    }
+}
+// NIFTI_TYPE_RGB24         128
+impl TypeTool<i8> for i8 {
+    fn to_data_type() -> i16 { 256 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: i8) {
+        writer.write_i8(d).unwrap();
+    }
+}
+impl TypeTool<u16> for u16 {
+    fn to_data_type() -> i16 { 512 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: u16) {
+        writer.write_u16::<B>(d).unwrap();
+    }
+}
+impl TypeTool<u32> for u32 {
+    fn to_data_type() -> i16 { 768 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: u32) {
+        writer.write_u32::<B>(d).unwrap();
+    }
+}
+impl TypeTool<i64> for i64 {
+    fn to_data_type() -> i16 { 1024 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: i64) {
+        writer.write_i64::<B>(d).unwrap();
+    }
+}
+impl TypeTool<u64> for u64 {
+    fn to_data_type() -> i16 { 1280 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: u64) {
+        writer.write_u64::<B>(d).unwrap();
+    }
+}
+// NIFTI_TYPE_FLOAT128     1536
+// NIFTI_TYPE_COMPLEX128   1792
+// NIFTI_TYPE_COMPLEX256   2048
+// NIFTI_TYPE_RGBA32       2304
+
+/// Write a nifti file (.nii or .nii.gz).
+///
+/// If a `reference` is given, it will be used to fill most of the header's fields. The voxels
+/// intensity will be subtracted by `scl_slope` and divided by `scl_inter`. If `reference` is not
+/// given, a default `NiftiHeader` will be built and written.
+///
+/// In all cases, the `dim`, `datatype` and `bitpix` fields will depend only on `data`, not on the
+/// header. In other words, the `datatype` defined in `reference` will be ignored.
+pub fn write_nifti<A, S, D>(
```

Comment:
Errors emerging from this function should be recoverable. In this case, changing the return type to `Result<(), NiftiError>` and raising errors with `?` should be enough.